File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/M._Золотая_середина/solution.cpp
#include <iostream>
#include <vector>
#include <cstdint>
#include <algorithm>

void solution() {
    /*
    Time Complexity O(log(min(A, B)))
    Memory Complexity O(1)
    */
    int n, m;
    std::cin >> n >> m;

    std::vector<int> A(n);
    std::vector<int> B(m);

    for (int i = 0; i < n; ++i) {
        std::cin >> A[i];
    }
    for (int i = 0; i < m; ++i) {
        std::cin >> B[i];
    }
    if (n > m) {
        std::swap(A, B);
        std::swap(n, m);
    }

    int left = 0, right = n, half = (n + m + 1) / 2;

    while (left <= right) {
        int middle = left + (right - left) / 2;
        int second_middle = half - middle;

        int Aleft = (middle > 0) ? A[middle - 1] : INT32_MIN;
        int Aright = (middle < n) ? A[middle] : INT32_MAX;

        int Bleft = (second_middle > 0) ? B[second_middle - 1] : INT32_MIN;
        int Bright = (second_middle < m) ? B[second_middle] : INT32_MAX;

        if (Aleft <= Bright && Bleft <= Aright) {
            if ((n + m) % 2 == 0) {
                std::cout << (std::max(Aleft, Bleft) + std::min(Aright, Bright)) / 2.0 << "\n";
            } else {
                std::cout << std::max(Aleft, Bleft) << "\n";
            }
            return;
        } else if (Aleft > Bright) {
            right = middle - 1;
        } else {
            left = middle + 1;
        }
    }
}

int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/3_sprint/M._Золотая_середина/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(log(min(A, B)))
    Memory Complexity O(1)
    """
    n: int = int(sys.stdin.readline().rstrip())
    m: int = int(sys.stdin.readline().rstrip())
    A: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    B: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    if n > m:
        A, B = B, A
        n, m = m, n
    half: int = (n + m) // 2
    left: int = 0
    right: int = n - 1
    while left < n:
        middle: int = left + (right - left) // 2
        second_middle: int = half - middle - 2
        A_left: int = A[middle] if middle >= 0 else float('-inf')
        A_right: int = A[middle + 1] if middle + 1 < n else float('inf')
        B_left: int = B[second_middle] if second_middle >= 0 else float('-inf')
        B_right: int = B[second_middle + 1] if second_middle + 1 < m else float('inf')
        if A_left <= B_right and B_left <= A_right:
            if (n + m) & 1:
                sys.stdout.write(f'{min(A_right, B_right)}\n')
                return
            else:
                median: float = (max(A_left, B_left) + min(A_right, B_right)) / 2
                if int(median) == median:
                    median = int(median)
                sys.stdout.write(f'{median}\n')
                return
        elif A_left > B_right:
            right = middle - 1
        else:
            left = middle + 1


if __name__ == '__main__':
    solution()